Ensure kernel launches are checked

r-barnes opened the pull request:
Summary:
Caffe2 and Torch currently does not have a consistent mechanism for determining if a kernel has launched successfully. The result is difficult-to-detect or silent errors. This diff provides functionality to fix that. Subsequent diffs on the stack fix the identified issues.

Kernel launch errors may arise if invalid launch parameters (number of blocks, number of threads, shared memory, or stream id) are specified incorrectly for the hardware or for other reasons. Interestingly, unless these launch errors are specifically checked for CUDA will silently fail and return garbage answers which can affect downstream computation. Therefore, catching launch errors is important.

Launches are currently checked by placing
```
AT_CUDA_CHECK(cudaGetLastError());
```
somewhere below the kernel launch. This is bad for two reasons.
1. The check may be performed at a site distant to the kernel launch, making debugging difficult.
2. The separation of the launch from the check means that it is difficult for humans and static analyzers to determine whether the check has taken place.

This diff defines a macro:
```
#define TORCH_CUDA_KERNEL_LAUNCH_CHECK() AT_CUDA_CHECK(cudaGetLastError())
```
which clearly indicates the check.

This diff also introduces a new test which analyzes code to identify kernel launches and determines whether the line immediately following the launch contains `TORCH_CUDA_KERNEL_LAUNCH_CHECK();`.

A search of the Caffe2 codebase identifies 104 instances of `AT_CUDA_CHECK(cudaGetLastError());` while the foregoing test identifies 1,467 launches which are not paired with a check. Visual inspection indicates that few of these are false positives, highlighting the need for some sort of static analysis system.

Test Plan:
The new test is run with:
```
buck run //caffe2/test:kernel_launch_error_check
```
And should be launched automatically with the other land tests. (TODO: Is it?)

The test is currently set up only to provide warnings but can later be adjusted to require checks.

Otherwise, I rely on the existing test frameworks to ensure that changes resulting from reorganizing existing launch checks don't cause regressions.

Differential Revision: D24309971

ezyang commented:
A few quick thoughts:
* `TORCH_CUDA_KERNEL_LAUNCH_CHECK`, don't use the `AT_` prefix, it's old
* The lint is currently being run as part of the entire test suite, but there is no reason to wait for PyTorch to be built to do this lint. This lint would be better incorporated as part of our fast lint checks so you can get feedback about it immediately, rather than having to wait for later.

r-barnes commented:
@ezyang : Could you point me to the fast lint checks? Also, should `TORCH_CUDA_CHECK()` and company be moved from `caffe2/torch/csrc/autograd/profiler_cuda.cpp` to `caffe2/aten/src/ATen/cuda/Exceptions.h`?

ngimel commented:
Fast lintchecks are in .github/workflows/lint.yml.

r-barnes commented:
@ezyang : This isn't working yet, but is this like what you had in mind?

ezyang commented on .github/workflows/lint.yml:
```diff
@@ -38,6 +38,10 @@ jobs:
           cd docs/cpp/source && ./check-doxygen.sh
+      - name: CUDA kernel launch check
+        run: |
+          set -eux
```
You sure this works? I feel like a pipefail is necessary as the pipe will suppress exit code from the first python

r-barnes replied:
This does correctly identify kernels without protection. I don't know if it will cause a lint failure because it's currently set up to always pass. Once the checks are in place, we can modify the lint check to be binding, so I suppose I'll discover at that point whether or not it integrates with the pipe correctly?

samestep replied:
This currently doesn't work in any case, since `tee` only uses stdout but `torch/testing/check_kernel_launches.py` only prints to stderr. Could you [replace `|` with `|&`](https://stackoverflow.com/a/363239/5044950)?

r-barnes replied:
Fixed.

samestep replied:
Oh sorry, I was referring to the pipe in the shell command, not the pipe indicating the start of the YAML string 😅

r-barnes replied:
Ooops! Fixed that.

samestep replied:
Thanks! Can you push the fix? I think that's the last thing blocking me from approving

ezyang commented:
@janeyx99 IIRC, you proposed more linting in one of the recent AI Infra syncs. This PR constitutes one way we could go about adding more lints, but it goes about doing it by adding a full on Python script per lint. However, it works for both OSS and fbcode. Are you interested in taking a look?

janeyx99 commented:
> @janeyx99 IIRC, you proposed more linting in one of the recent AI Infra syncs. This PR constitutes one way we could go about adding more lints, but it goes about doing it by adding a full on Python script per lint. However, it works for both OSS and fbcode. Are you interested in taking a look?

Yes, this looks to be the way we had in mind as well. I'd be down to take a look! Currently @samestep has been looking into implementing this, so he may also have more input about this.

samestep commented:
@ezyang @janeyx99 Oh interesting, so this PR is just uses a standalone Python script to lint? That could work I think; I was planning to just write an in-tree [Flake8 plugin](https://flake8.pycqa.org/en/latest/plugin-development/index.html) and integrate that into our existing Flake8 usage in OSS CI. Does fbcode also run the Flake8 checks, or no?

ezyang commented:
one benefit of making a flake8 plugin is that you would be able to use annotations to annotate the diff directly with the lint information without any extra work. Without it, this has to be reimplemented from scratch for the custom lint script.

ezyang commented:
flake8 is not a very good choice for the lint here, though, which is operating on C++ code.

r-barnes commented:
@samestep, @janeyx99 , @ezyang: Sorry about the delay. I've lost the thread of the conversation. @ezyang's suggestion to integrate this with flake to get annotations sounds compelling, but I'm not sure if that would still work as an internal test.

I'm also unfamiliar with how to build flake plugins and such, so I'm not sure if that's a good use of my individual time, though I'd be happy to collaborate with any of you on a diff to move a similar functionality into flake.

I think I'd like to land this so I can get some bootcampers working on improving the kernel safety and moving the codebase toward the point where an enforcing check can be used, irrespective of how we accomplish that.

ngimel commented:
Given that very few OSS PRs are actually adding kernel launches, I agree that blocking on flake plugin to get inline annotations does not make sense, just test error should be enough.

ezyang commented:
Sorry, to clarify, I was simply trying to identify the benefits and downsides to modeling this as a flake8 plugin. I don't have a particular preference one way or another.

However, to be more explicit, my preferred outcome is to get @samestep or @janeyx99 to approve this PR, since your team is likely to become "owners" of infrastructure like this.

samestep commented:
> Subsequent diffs on the stack fix the identified issues.

Could you clarify the meaning of this sentence @r-barnes? I only see one diff in the stack.

r-barnes commented:
Thanks for the clarification, @ezyang 

@samestep : Sorry about that, I'm referring to D24334303 which was previously on a stack with this commit. I've since moved it off in hopes of accelerating adding the launch checks, since adding the checks and adding the static analysis to find them can proceed independently.

samestep commented on test/test_kernel_launch_checks.py:
```diff
@@ -0,0 +1,38 @@
+from torch.testing._internal.common_utils import TestCase
```
Could you modify this line to also import `run_tests`, and then also add the following to the bottom of the file?
```python
if __name__ == '__main__':
    run_tests()
```
This would allow your new test file to be run using `python test/test_kernel_launch_checks.py` similar to existing `test/*.py` files.

r-barnes replied:
Fixed.

samestep commented on torch/testing/__init__.py:
```diff
@@ -6,6 +6,7 @@
 import random
 import math
 from typing import cast, List, Optional, Tuple, Union
+from .check_kernel_launches import check_cuda_kernel_launches, check_code_for_cuda_kernel_launches
```
Could you clarify the purpose of this line? I don't see any similar ones nearby and the imported functions don't seem to be used here, so it appears to be redundant.

r-barnes replied:
This exports `check_cuda_kernel_launches` and `check_code_for_cuda_kernel_launches` into the testing namespace so they can be used by the OSS linter.

samestep replied:
Hmm ok; could you give an example of a command that breaks when you remove that line?

r-barnes replied:
Line 2 of  test/test_kernel_launch_checks.py

samestep replied:
Oh right! sorry, not sure how I missed that

samestep commented on test/test_kernel_launch_checks.py:
```diff
@@ -0,0 +1,38 @@
+        check_cuda_kernel_launches()
+        # TODO: Enable this after warning messages have been dealt with.
+        self.assertTrue(True)
+        # self.assertTrue(check_cuda_kernel_launches() == 0)
```
Just to clarify, is this TODO meant to be resolved by [D24334303](https://www.internalfb.com/diff/D24334303)?

r-barnes replied:
D24334303 only fixes a few of the 1700 some kernels that need to be adjusted, so it will take a few more diffs, but ultimately, yes.

r-barnes committed:
Ensure kernel launches are checked (#46474)

Summary:
Caffe2 and Torch currently does not have a consistent mechanism for determining if a kernel has launched successfully. The result is difficult-to-detect or silent errors. This diff provides functionality to fix that. Subsequent diffs on the stack fix the identified issues.

Kernel launch errors may arise if invalid launch parameters (number of blocks, number of threads, shared memory, or stream id) are specified incorrectly for the hardware or for other reasons. Interestingly, unless these launch errors are specifically checked for CUDA will silently fail and return garbage answers which can affect downstream computation. Therefore, catching launch errors is important.

Launches are currently checked by placing
```
AT_CUDA_CHECK(cudaGetLastError());
```
somewhere below the kernel launch. This is bad for two reasons.
1. The check may be performed at a site distant to the kernel launch, making debugging difficult.
2. The separation of the launch from the check means that it is difficult for humans and static analyzers to determine whether the check has taken place.

This diff defines a macro:
```
#define TORCH_CUDA_KERNEL_LAUNCH_CHECK() AT_CUDA_CHECK(cudaGetLastError())
```
which clearly indicates the check.

This diff also introduces a new test which analyzes code to identify kernel launches and determines whether the line immediately following the launch contains `TORCH_CUDA_KERNEL_LAUNCH_CHECK();`.

A search of the Caffe2 codebase identifies 104 instances of `AT_CUDA_CHECK(cudaGetLastError());` while the foregoing test identifies 1,467 launches which are not paired with a check. Visual inspection indicates that few of these are false positives, highlighting the need for some sort of static analysis system.

Pull Request resolved: https://github.com/pytorch/pytorch/pull/46474

Test Plan:
The new test is run with:
```
buck test //caffe2/test:kernel_launch_checks -- --print-passing-details
```
And should be launched automatically with the other land tests. (TODO: Is it?)

The test is currently set up only to provide warnings but can later be adjusted to require checks.

Otherwise, I rely on the existing test frameworks to ensure that changes resulting from reorganizing existing launch checks don't cause regressions.

Reviewed By: ngimel

Differential Revision: D24309971

Pulled By: r-barnes

fbshipit-source-id: e1a90edd93c64d54cc64f1ba28f48fd6186a0a67
